pkg/infra/baseagent.go

```diff
@@ -93,8 +95,15 @@ func (a *Base) setHP(newHP int) {
 	a.hp = newHP
 }
 
+// Modeled as a first order system step answer (see documentation for more information)
 func (a *Base) updateHP(foodTaken float64) {
-	a.hp = int(math.Min(100, float64(a.hp)+foodTaken))
+
+	if a.hp >= a.tower.healthInfo.WeakLevel {
+		a.hp = a.hp + int(a.tower.healthInfo.Width*(1-math.Pow(math.E, -foodTaken/a.tower.healthInfo.Tau)))
+	} else {
+		a.hp = int(math.Min(float64(a.tower.healthInfo.HPReqCToW), float64(a.hp+int(a.tower.healthInfo.Width*(1-math.Pow(math.E, -foodTaken/a.tower.healthInfo.Tau))))))
```

Comment:
```suggestion
        hpChange := a.tower.healthInfo.Width*(1-math.Pow(math.E, -foodTaken/a.tower.healthInfo.Tau))
	if a.hp >= a.tower.healthInfo.WeakLevel {
		a.hp = a.hp + int(hpChange)
	} else {
		a.hp = int(math.Min(float64(a.tower.healthInfo.HPReqCToW), float64(a.hp)+hpChange))
```

Reply:
Actually... should this be max or min? I'm not sure what it's doing.

Reply:
`a.hp = a.hp + int(math.Min(float64(a.tower.healthInfo.HPReqCToW), float64(hpChange))`

Not sure if this is what you meant to have here?? This prevents the agent from going into the weak state just to be put back into it. And allows the agent to pass the check in hpDecay of `if agent.hp >= t.healthInfo.HPCritical+t.healthInfo.HPReqCToW` which will then put them into the weak state correctly.

Reply:
Rhys is right. There might be other ways to do it but I think it should work quite nicely